File: internal/repository/registration_repository.go
package repository

import (
	"context"
	"database/sql"
	"fmt"

	"github.com/Homeria/baeum-maru/internal/database"
	"github.com/Homeria/baeum-maru/internal/domain"
)

type RegistrationRepository struct {
	db *sql.DB
}

type CreateRegistrationParams struct {
	MemberID   int64
	OfferingID int64
}

func NewRegistrationRepository(db *sql.DB) *RegistrationRepository {
	return &RegistrationRepository{db: db}
}

func (r *RegistrationRepository) Create(ctx context.Context, params CreateRegistrationParams) (domain.Registration, error) {
	reactivated, err := r.reactivateCancelled(ctx, params)
	if err != nil {
		return domain.Registration{}, err
	}
	if reactivated {
		return r.GetByMemberAndOffering(ctx, params.MemberID, params.OfferingID)
	}

	result, err := r.db.ExecContext(ctx, `
INSERT INTO registrations (member_id, offering_id, status)
VALUES (?, ?, 'applied');
`, params.MemberID, params.OfferingID)
	if err != nil {
		return domain.Registration{}, fmt.Errorf("insert registration: %w", err)
	}

	id, err := result.LastInsertId()
	if err != nil {
		return domain.Registration{}, fmt.Errorf("read registration id: %w", err)
	}
	return r.Get(ctx, id)
}

func (r *RegistrationRepository) reactivateCancelled(ctx context.Context, params CreateRegistrationParams) (bool, error) {
	result, err := r.db.ExecContext(ctx, `
UPDATE registrations
SET status = 'applied',
    cancelled_at = NULL,
    updated_at = CURRENT_TIMESTAMP
WHERE member_id = ? AND offering_id = ? AND status = 'cancelled';
`, params.MemberID, params.OfferingID)
	if err != nil {
		return false, fmt.Errorf("reactivate cancelled registration: %w", err)
	}

	affected, err := result.RowsAffected()
	if err != nil {
		return false, fmt.Errorf("read reactivated registration rows: %w", err)
	}
	return affected > 0, nil
}

func (r *RegistrationRepository) Cancel(ctx context.Context, id int64) (domain.Registration, error) {
	change, err := r.CancelAndPromote(ctx, id)
	if err != nil {
		return domain.Registration{}, err
	}
	return change.Registration, nil
}

func (r *RegistrationRepository) Confirm(ctx context.Context, id int64) (domain.Registration, error) {
	var confirmed domain.Registration
	err := database.WithTx(ctx, r.db, func(tx *sql.Tx) error {
		if err := updateRegistrationStatus(ctx, tx, id, "selected", "confirmed", "registration_confirmed"); err != nil {
			return err
		}

		registration, err := r.getWithTx(ctx, tx, id)
		if err != nil {
			return err
		}
		confirmed = registration
		return nil
	})
	if err != nil {
		return domain.Registration{}, err
	}
	return confirmed, nil
}

func (r *RegistrationRepository) CancelAndPromote(ctx context.Context, id int64) (domain.RegistrationStatusChange, error) {
	var change domain.RegistrationStatusChange
	err := database.WithTx(ctx, r.db, func(tx *sql.Tx) error {
		var fromStatus string
		var offeringID int64
		if err := tx.QueryRowContext(ctx, "SELECT status, offering_id FROM registrations WHERE id = ?;", id).Scan(&fromStatus, &offeringID); err != nil {
			return fmt.Errorf("read registration for cancellation: %w", err)
		}
		if fromStatus == "cancelled" {
			return sql.ErrNoRows
		}

		result, err := tx.ExecContext(ctx, `
UPDATE registrations
SET status = 'cancelled',
    cancelled_at = CURRENT_TIMESTAMP,
    updated_at = CURRENT_TIMESTAMP
WHERE id = ? AND status != 'cancelled';
`, id)
		if err != nil {
			return fmt.Errorf("cancel registration %d: %w", id, err)
		}

		affected, err := result.RowsAffected()
		if err != nil {
			return fmt.Errorf("read affected registration rows: %w", err)
		}
		if affected == 0 {
			return sql.ErrNoRows
		}
		if err := insertRegistrationStatusHistory(ctx, tx, id, fromStatus, "cancelled", "registration_cancelled"); err != nil {
			return err
		}

		cancelled, err := r.getWithTx(ctx, tx, id)
		if err != nil {
			return err
		}
		change.Registration = cancelled

		if fromStatus == "selected" || fromStatus == "confirmed" {
			promoted, err := promoteNextWaitlisted(ctx, tx, offeringID, id)
			if err != nil {
				return err
			}
			change.Promoted = promoted
		}
		return nil
	})
	if err != nil {
		return domain.RegistrationStatusChange{}, err
	}
	return change, nil
}

func (r *RegistrationRepository) Get(ctx context.Context, id int64) (domain.Registration, error) {
	row := r.db.QueryRowContext(ctx, registrationSelectSQL()+`
WHERE r.id = ?;
`, id)
	return scanRegistration(row)
}

func (r *RegistrationRepository) getWithTx(ctx context.Context, tx *sql.Tx, id int64) (domain.Registration, error) {
	row := tx.QueryRowContext(ctx, registrationSelectSQL()+`
WHERE r.id = ?;
`, id)
	return scanRegistration(row)
}

func (r *RegistrationRepository) GetByMemberAndOffering(ctx context.Context, memberID int64, offeringID int64) (domain.Registration, error) {
	row := r.db.QueryRowContext(ctx, registrationSelectSQL()+`
WHERE r.member_id = ? AND r.offering_id = ?;
`, memberID, offeringID)
	return scanRegistration(row)
}

func (r *RegistrationRepository) ListByMember(ctx context.Context, memberID int64) ([]domain.Registration, error) {
	rows, err := r.db.QueryContext(ctx, registrationSelectSQL()+`
WHERE r.member_id = ?
ORDER BY r.created_at DESC, r.id DESC;
`, memberID)
	if err != nil {
		return nil, fmt.Errorf("list registrations by member: %w", err)
	}
	defer rows.Close()

	return scanRegistrations(rows)
}

func (r *RegistrationRepository) ListByOffering(ctx context.Context, offeringID int64) ([]domain.Registration, error) {
	rows, err := r.db.QueryContext(ctx, registrationSelectSQL()+`
WHERE r.offering_id = ?
ORDER BY m.name, r.id;
`, offeringID)
	if err != nil {
		return nil, fmt.Errorf("list registrations by offering: %w", err)
	}
	defer rows.Close()

	return scanRegistrations(rows)
}

func (r *RegistrationRepository) ListRecent(ctx context.Context, limit int) ([]domain.Registration, error) {
	if limit <= 0 {
		limit = 100
	}
	if limit > 10000 {
		limit = 10000
	}

	rows, err := r.db.QueryContext(ctx, registrationSelectSQL()+`
ORDER BY r.created_at DESC, r.id DESC
LIMIT ?;
`, limit)
	if err != nil {
		return nil, fmt.Errorf("list recent registrations: %w", err)
	}
	defer rows.Close()

	return scanRegistrations(rows)
}

func (r *RegistrationRepository) ListActiveRuleItemsByMember(ctx context.Context, memberID int64) ([]domain.RegistrationRuleItem, error) {
	rows, err := r.db.QueryContext(ctx, `
SELECT r.id,
       r.member_id,
       r.offering_id,
       co.term_id,
       r.status,
       CASE WHEN cs.weekday = 7 THEN 0 ELSE cs.weekday END,
       ts.start_time,
       ts.end_time
FROM registrations r
JOIN course_offerings co ON co.id = r.offering_id
JOIN course_schedules cs ON cs.offering_id = co.id
JOIN time_slots ts ON ts.id = cs.time_slot_id
WHERE r.member_id = ?
  AND r.status IN ('applied', 'selected', 'waitlisted', 'confirmed')
ORDER BY r.id;
`, memberID)
	if err != nil {
		return nil, fmt.Errorf("list active registration rule items: %w", err)
	}
	defer rows.Close()

	var items []domain.RegistrationRuleItem
	for rows.Next() {
		var item domain.RegistrationRuleItem
		if err := rows.Scan(&item.ID, &item.MemberID, &item.OfferingID, &item.TermID, &item.Status, &item.Weekday, &item.StartTime, &item.EndTime); err != nil {
			return nil, fmt.Errorf("scan registration rule item: %w", err)
		}
		items = append(items, item)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate registration rule items: %w", err)
	}
	return items, nil
}

func registrationSelectSQL() string {
	return `
SELECT r.id,
       m.id,
       m.name,
       COALESCE(m.member_no, ''),
       co.id,
       co.display_name,
       t.name,
       r.status,
       r.created_at,
       r.updated_at,
       COALESCE(r.cancelled_at, '')
FROM registrations r
JOIN members m ON m.id = r.member_id
JOIN course_offerings co ON co.id = r.offering_id
JOIN terms t ON t.id = co.term_id
`
}

func scanRegistrations(rows *sql.Rows) ([]domain.Registration, error) {
	var registrations []domain.Registration
	for rows.Next() {
		registration, err := scanRegistration(rows)
		if err != nil {
			return nil, err
		}
		registrations = append(registrations, registration)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate registrations: %w", err)
	}
	return registrations, nil
}

func scanRegistration(scanner interface{ Scan(dest ...any) error }) (domain.Registration, error) {
	var registration domain.Registration
	if err := scanner.Scan(
		&registration.ID,
		&registration.MemberID,
		&registration.MemberName,
		&registration.MemberNo,
		&registration.OfferingID,
		&registration.CourseTitle,
		&registration.TermName,
		&registration.Status,
		&registration.CreatedAt,
		&registration.UpdatedAt,
		&registration.CancelledAt,
	); err != nil {
		return domain.Registration{}, fmt.Errorf("scan registration: %w", err)
	}
	return registration, nil
}

func updateRegistrationStatus(ctx context.Context, tx *sql.Tx, id int64, fromStatus string, toStatus string, reason string) error {
	result, err := tx.ExecContext(ctx, `
UPDATE registrations
SET status = ?,
    cancelled_at = CASE WHEN ? = 'cancelled' THEN CURRENT_TIMESTAMP ELSE cancelled_at END,
    updated_at = CURRENT_TIMESTAMP
WHERE id = ? AND status = ?;
`, toStatus, toStatus, id, fromStatus)
	if err != nil {
		return fmt.Errorf("update registration status: %w", err)
	}
	affected, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("read affected registration rows: %w", err)
	}
	if affected == 0 {
		return sql.ErrNoRows
	}
	return insertRegistrationStatusHistory(ctx, tx, id, fromStatus, toStatus, reason)
}

func promoteNextWaitlisted(ctx context.Context, tx *sql.Tx, offeringID int64, cancelledRegistrationID int64) (*domain.Registration, error) {
	var promotedID int64
	err := tx.QueryRowContext(ctx, `
SELECT r.id
FROM registrations r
LEFT JOIN lottery_results lr ON lr.registration_id = r.id
WHERE r.offering_id = ?
  AND r.status = 'waitlisted'
ORDER BY COALESCE(lr.result_order, r.id), r.created_at, r.id
LIMIT 1;
`, offeringID).Scan(&promotedID)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("find next waitlisted registration: %w", err)
	}
	if err := updateRegistrationStatus(ctx, tx, promotedID, "waitlisted", "selected", fmt.Sprintf("waitlist_promoted_after_cancel:%d", cancelledRegistrationID)); err != nil {
		return nil, err
	}

	row := tx.QueryRowContext(ctx, registrationSelectSQL()+`
WHERE r.id = ?;
`, promotedID)
	promoted, err := scanRegistration(row)
	if err != nil {
		return nil, err
	}
	return &promoted, nil
}

func insertRegistrationStatusHistory(ctx context.Context, tx *sql.Tx, registrationID int64, fromStatus string, toStatus string, reason string) error {
	if _, err := tx.ExecContext(ctx, `
INSERT INTO registration_status_history (registration_id, from_status, to_status, reason)
VALUES (?, ?, ?, ?);
`, registrationID, fromStatus, toStatus, reason); err != nil {
		return fmt.Errorf("insert registration status history: %w", err)
	}
	return nil
}

File: backend/app/repositories/registration_repository.py
"""강좌 신청과 상태 이력 데이터를 다루는 repository."""

import sqlite3
from typing import Any

from app.core.exceptions import ConflictError
from app.db.database import get_db_connection

# 취소·탈락이 아닌 유효한 신청 상태
_ACTIVE = ("applied", "selected", "waitlisted", "confirmed")
_ACTIVE_SQL = "('applied', 'selected', 'waitlisted', 'confirmed')"


def _history(
    conn: sqlite3.Connection,
    registration_id: int,
    from_status: str | None,
    to_status: str,
    reason: str | None = None,
    *,
    actor_operator_id: int | None = None,
    actor_display_name: str | None = None,
) -> None:
    actor_kind = "operator" if actor_operator_id is not None else "system"
    conn.execute(
        """
        INSERT INTO registration_status_history
            (registration_id, from_status, to_status, reason,
             actor_kind, actor_operator_id, actor_display_name)
        VALUES (?, ?, ?, ?, ?, ?, ?)
        """,
        (
            registration_id,
            from_status,
            to_status,
            reason,
            actor_kind,
            actor_operator_id,
            actor_display_name,
        ),
    )


# --- Read ---


def get_registration(registration_id: int) -> dict[str, Any] | None:
    with get_db_connection() as conn:
        row = conn.execute(
            "SELECT * FROM registrations WHERE id = ?", (registration_id,)
        ).fetchone()
    return dict(row) if row is not None else None


def list_registrations(
    *,
    member_id: int | None = None,
    offering_id: int | None = None,
    term_id: int | None = None,
    status: str | None = None,
) -> list[dict[str, Any]]:
    query = "SELECT * FROM registrations"
    conditions: list[str] = []
    params: list[Any] = []
    if member_id is not None:
        conditions.append("member_id = ?")
        params.append(member_id)
    if offering_id is not None:
        conditions.append("offering_id = ?")
        params.append(offering_id)
    if status is not None:
        conditions.append("status = ?")
        params.append(status)
    if term_id is not None:
        conditions.append("offering_id IN (SELECT id FROM course_offerings WHERE term_id = ?)")
        params.append(term_id)
    if conditions:
        query += " WHERE " + " AND ".join(conditions)
    query += " ORDER BY id"
    with get_db_connection() as conn:
        rows = conn.execute(query, params).fetchall()
    return [dict(row) for row in rows]


def list_history(registration_id: int) -> list[dict[str, Any]]:
    with get_db_connection() as conn:
        rows = conn.execute(
            "SELECT * FROM registration_status_history WHERE registration_id = ? "
            "ORDER BY changed_at, id",
            (registration_id,),
        ).fetchall()
    return [dict(row) for row in rows]


def count_active_in_term(member_id: int, term_id: int) -> int:
    with get_db_connection() as conn:
        row = conn.execute(
            f"""
            SELECT COUNT(*) FROM registrations r
            JOIN course_offerings o ON o.id = r.offering_id
            WHERE r.member_id = ? AND o.term_id = ? AND r.status IN {_ACTIVE_SQL}
            """,
            (member_id, term_id),
        ).fetchone()
    return int(row[0])


def get_member_active_slots(member_id: int) -> list[tuple[int, int]]:
    with get_db_connection() as conn:
        rows = conn.execute(
            f"""
            SELECT cs.weekday, cs.time_slot_id FROM course_schedules cs
            JOIN registrations r ON r.offering_id = cs.offering_id
            WHERE r.member_id = ? AND r.status IN {_ACTIVE_SQL}
            """,
            (member_id,),
        ).fetchall()
    return [(int(r[0]), int(r[1])) for r in rows]


def get_offering_slots(offering_id: int) -> list[tuple[int, int]]:
    with get_db_connection() as conn:
        rows = conn.execute(
            "SELECT weekday, time_slot_id FROM course_schedules WHERE offering_id = ?",
            (offering_id,),
        ).fetchall()
    return [(int(r[0]), int(r[1])) for r in rows]


# --- Write (원자적) ---


def apply_registrations(
    member_id: int,
    offering_ids: list[int],
    *,
    actor_operator_id: int | None = None,
    actor_display_name: str | None = None,
) -> list[dict[str, Any]]:
    """여러 강좌 신청을 한 transaction으로 처리한다(전부 아니면 전무).

    이미 취소/탈락한 신청은 다시 applied로 되살리고, 활성 신청이면 전체를 취소한다.
    """
    with get_db_connection() as conn:
        try:
            ids: list[int] = []
            for offering_id in offering_ids:
                existing = conn.execute(
                    "SELECT id, status FROM registrations WHERE member_id = ? AND offering_id = ?",
                    (member_id, offering_id),
                ).fetchone()
                if existing is None:
                    cursor = conn.execute(
                        "INSERT INTO registrations (member_id, offering_id, status) "
                        "VALUES (?, ?, 'applied')",
                        (member_id, offering_id),
                    )
                    rid = int(cursor.lastrowid)  # type: ignore[arg-type]
                    _history(
                        conn,
                        rid,
                        None,
                        "applied",
                        actor_operator_id=actor_operator_id,
                        actor_display_name=actor_display_name,
                    )
                elif existing["status"] in ("cancelled", "rejected"):
                    rid = int(existing["id"])
                    conn.execute(
                        "UPDATE registrations SET status = 'applied', waitlist_order = NULL, "
                        "updated_at = CURRENT_TIMESTAMP WHERE id = ?",
                        (rid,),
                    )
                    _history(
                        conn,
                        rid,
                        str(existing["status"]),
                        "applied",
                        actor_operator_id=actor_operator_id,
                        actor_display_name=actor_display_name,
                    )
                else:
                    raise ConflictError("already_applied", "이미 신청한 강좌가 있습니다.")
                ids.append(rid)
            conn.commit()
        except sqlite3.IntegrityError as error:
            conn.rollback()
            raise ConflictError(
                "registration_conflict", "신청 처리 중 충돌이 발생했습니다."
            ) from error
        except BaseException:
            conn.rollback()
            raise
        return [
            dict(conn.execute("SELECT * FROM registrations WHERE id = ?", (i,)).fetchone())
            for i in ids
        ]


def cancel_registration(
    registration_id: int,
    reason: str | None,
    *,
    actor_operator_id: int | None = None,
    actor_display_name: str | None = None,
) -> dict[str, Any] | None:
    """신청을 취소하고, 당첨자였다면 대기 순번대로 승계한다(한 transaction)."""
    with get_db_connection() as conn:
        reg = conn.execute(
            "SELECT * FROM registrations WHERE id = ?", (registration_id,)
        ).fetchone()
        if reg is None:
            return None
        if reg["status"] == "cancelled":
            return dict(reg)
        try:
            conn.execute(
                "UPDATE registrations SET status = 'cancelled', updated_at = CURRENT_TIMESTAMP "
                "WHERE id = ?",
                (registration_id,),
            )
            _history(
                conn,
                registration_id,
                str(reg["status"]),
                "cancelled",
                reason,
                actor_operator_id=actor_operator_id,
                actor_display_name=actor_display_name,
            )
            if reg["status"] == "selected":
                nxt = conn.execute(
                    "SELECT id FROM registrations WHERE offering_id = ? AND status = 'waitlisted' "
                    "ORDER BY waitlist_order LIMIT 1",
                    (reg["offering_id"],),
                ).fetchone()
                if nxt is not None:
                    conn.execute(
                        "UPDATE registrations SET status = 'selected', "
                        "updated_at = CURRENT_TIMESTAMP WHERE id = ?",
                        (int(nxt["id"]),),
                    )
                    _history(
                        conn,
                        int(nxt["id"]),
                        "waitlisted",
                        "selected",
                        "대기 승계",
                        actor_operator_id=actor_operator_id,
                        actor_display_name=actor_display_name,
                    )
            conn.commit()
        except BaseException:
            conn.rollback()
            raise
        return dict(
            conn.execute("SELECT * FROM registrations WHERE id = ?", (registration_id,)).fetchone()
        )
